beep/utils/waveform.py

```diff
@@ -56,3 +58,246 @@ def convert_velocity_to_power_waveform(waveform_file, velocity_units):
     power = -power * df.v  # positive power = charge
 
     return pd.DataFrame({"time": df.t, "power": power})
+
+
+class RapidChargeWave:
+    """
+       Object to produce charging waveforms, with direct comparison between stepwise and
+       smooth charging waveforms. Algorithms written by Patrick Asinger.
+
+    """
+    def __init__(self,
+                 final_c_rate,
+                 soc_initial,
+                 soc_final,
+                 max_c_rate,
+                 min_c_rate
+                 ):
+        """
+        Args:
+            final_c_rate (float): charging rate for the final step
+            soc_initial (float): estimated starting soc for the fast charging portion of the cycle
+            soc_final (float): estimated soc to end the fast charging portion of the cycle
+            max_c_rate (float): maximum charging rate for the charge
+            min_c_rate (float): minimum charging rate for the charge
+
+        """
+        self.final_c_rate = final_c_rate
+        self.soc_i = soc_initial
+        self.soc_f = soc_final
+        self.max_c_rate = max_c_rate
+        self.min_c_rate = min_c_rate
+        self.soc_points = 1000
+
+    def get_currents_with_uniform_time_basis(self, charging_c_rates, mesh_points):
+        """
+        Function to re-interpolate all of the current values to a uniform time basis, with each time step 1 sec
+
+        Args:
+            charging_c_rates (list): c-rates for each of the charging steps. Each step is assumed to be an equal
+                SOC portion of the charge, and the length of the list just needs to be at least 1
+            mesh_points (list): soc values for beginning and end of each of the charging windows
+
+        returns
+        np.array: array with the adjusted smooth current as a function of uniform time basis
+        np.array: array with the multistep current as a function of the uniform time basis
+        np.array: array with the corresponding uniformly spaced time values
+        """
+        current_smooth, time_smooth, current_multistep, time_multistep = \
+            self.get_input_currents_both_to_final_soc(charging_c_rates, mesh_points)
+        end_time_smooth = np.int(np.round(np.max(time_smooth), 0))
```

Comment:
Would it help to assert that `mesh_points` (soc_list) is monotonically increasing?

Reply:
Good idea, adding that in